crypto/fipsmodule/rand/ctrdrbg.c

```diff
@@ -12,21 +12,207 @@
  * OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
  * CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE. */
 
+#include <openssl/aes.h>
 #include <openssl/ctrdrbg.h>
 
 #include <openssl/type_check.h>
 #include <openssl/mem.h>
 
 #include "internal.h"
 #include "../cipher/internal.h"
-
+#include "../delocate.h"
+#include "../modes/internal.h"
+#include "../service_indicator/internal.h"
 
 // Section references in this file refer to SP 800-90Ar1:
 // http://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-90Ar1.pdf
 
 // See table 3.
 static const uint64_t kMaxReseedCount = UINT64_C(1) << 48;
 
+// kInitMask is the result of encrypting blocks with big-endian value 1, 2
+// and 3 with the all-zero AES-256 key.
+static const uint8_t kInitMask[CTR_DRBG_ENTROPY_LEN] = {
+    0x53, 0x0f, 0x8a, 0xfb, 0xc7, 0x45, 0x36, 0xb9, 0xa9, 0x63, 0xb4, 0xf1,
+    0xc4, 0xcb, 0x73, 0x8b, 0xce, 0xa7, 0x40, 0x3d, 0x4d, 0x60, 0x6b, 0x6e,
+    0x07, 0x4e, 0xc5, 0xd3, 0xba, 0xf3, 0x9d, 0x18, 0x72, 0x60, 0x03, 0xca,
+    0x37, 0xa6, 0x2a, 0x74, 0xd1, 0xa2, 0xf5, 0x8e, 0x75, 0x06, 0x35, 0x8e,
+};
+
+struct generation_nonce {
+  uint8_t nonce[CTR_DRBG_NONCE_LEN];
+};
+
+DEFINE_BSS_GET(AES_KEY, bcc_key)
+DEFINE_BSS_GET(struct generation_nonce, generation_nonce)
+DEFINE_STATIC_ONCE(bcc_init_once)
+DEFINE_STATIC_MUTEX(generation_nonce_lock)
+
+// Implementation of derivation function assumes CTR_DRBG_ENTROPY_LEN is 48.
+// That is, assumes the number of output bytes is 48. In addition, value must be
+// a multiple of the AES block size.
+OPENSSL_STATIC_ASSERT(CTR_DRBG_ENTROPY_LEN == 48, ctr_drbg_df_assumes_ctr_drbg_entropy_len_is_equal_to_48)
+OPENSSL_STATIC_ASSERT(CTR_DRBG_ENTROPY_LEN % AES_BLOCK_SIZE == 0,
+                      not_a_multiple_of_AES_block_size)
+
+static void init_df_bcc_once(void) {
```

Comment:
minor nitpick: i usually consider "once" functions to themselves be idempotent, but it looks like idempotency here is granted by `CRYPTO_once` and not this function. suggest dropping the `_once` suffix